src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+
+    /// Static memory manager to manage the memory available for blocks.
+    static MEMORY_MANAGER: RefCell<MemoryManager<Memory>> = RefCell::new(MemoryManager::init(log_memory()));
+
+    /// Append-only list of encoded blocks stored in stable memory.
+    static LOG: RefCell<StableLog> = with_memory_manager(|memory_manager| {
+        RefCell::new(Log::init(memory_manager.get(LOG_INDEX_MEMORY_ID), memory_manager.get(LOG_DATA_MEMORY_ID)).expect("failed to initialize stable log"))
+    });
+
+    /// Index to efficiently filter entries by user number.
+    static USER_INDEX: RefCell<UserIndex> = with_memory_manager(|memory_manager| {
+        RefCell::new(StableBTreeMap::new(memory_manager.get(USER_INDEX_MEMORY_ID), USER_INDEX_KEY_LENGTH as u32, 0))
+    });
+}
+
+/// Creates a memory region for the configuration stable cell.
+fn config_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 0..1)
+}
+
+/// Creates a memory region for the append-only block list.
+fn log_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 1..MAX_WASM_PAGES)
+}
+
+/// A helper function to access the configuration.
+fn with_config<R>(f: impl FnOnce(&ArchiveConfig) -> R) -> R {
+    CONFIG.with(|cell| f(cell.borrow().get()))
+}
+
+/// A helper function to access the memory manager.
+fn with_memory_manager<R>(f: impl FnOnce(&MemoryManager<Memory>) -> R) -> R {
+    MEMORY_MANAGER.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the log.
+fn with_log<R>(f: impl FnOnce(&StableLog) -> R) -> R {
+    LOG.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the user index.
+fn with_user_index<R>(f: impl FnOnce(&mut UserIndex) -> R) -> R {
+    USER_INDEX.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+/// Configuration of the archive canister.
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct ArchiveConfig {
+    /// This canister will accept entries only from this principal.
+    ii_canister: Principal,
+    /// The maximum number of entries returned in a single read canister call.
+    max_entries_per_call: u16,
+    /// Timestamp of the last install / upgrade of this canister.
+    last_upgrade_timestamp: Timestamp,
+}
+
+impl Storable for ArchiveConfig {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let buf = candid::encode_one(&self).expect("failed to encode log config");
+        Cow::Owned(buf)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode log options")
+    }
+}
+
+impl Default for ArchiveConfig {
+    fn default() -> Self {
+        Self {
+            ii_canister: Principal::from_text("rdmx6-jaaaa-aaaaa-aaadq-cai").unwrap(),
+            max_entries_per_call: 1000,
+            last_upgrade_timestamp: 0,
+        }
+    }
+}
+
+/// Index key for the user index.
+#[derive(Debug)]
+struct UserIndexKey {
+    user_number: UserNumber,
+    timestamp: Timestamp,
```

Comment:
I'm a little confused; why is there a timestamp in the key? _EDIT: see my question around the `UserIndex` type instead_

Reply:
This is there to allow finding entries for an anchor based on the `timestamp` as well (which might be useful for anchors with a lot of activity or just after using it for a long time).

Added a clarifying comment.